roles/DCOS.master/tasks/dcos_upgrade.yml

```diff
@@ -9,6 +14,7 @@
     url: "{{ dcos['config']['bootstrap_url'] }}/{{ dcos_version_specifier }}/genconf/serve/upgrade_from_{{ ansible_local.dcos_installation['version'] }}/latest/dcos_node_upgrade.sh"
     dest: "{{ path_base_tmp }}/{{ dcos_version_specifier }}/upgrade_from_{{ ansible_local.dcos_installation['version'] }}/"
     mode: 0740
+    force: yes
```

Comment:
I really like the idea of purging to keep nodes "clean" but the current purge option imo isn't needed as it would not actually purge the node completely of the older upgrades but would only purge the nodes of config style upgrades. This is because we are specifying current dcos version and upgrade from as directory names which would in turn be recreated in the next task and then ignored on next version upgrade.

Could we look into a way to purge all upgrades? such as purging entire `path_base_tmp`? This would ensure purging all older upgrades.

To ensure config change upgrades, we should then use `force: yes` and full `dest` path like so:

```
- name: "Upgrade: Download dcos_node_upgrade.sh"
  get_url:
    url: "{{ dcos['config']['bootstrap_url'] }}/{{ dcos_version_specifier }}/genconf/serve/upgrade_from_{{ ansible_local.dcos_installation['version'] }}/latest/dcos_node_upgrade.sh"
    dest: "{{ path_base_tmp }}/{{ dcos_version_specifier }}/upgrade_from_{{ ansible_local.dcos_installation['version'] }}/dcos_node_upgrade.sh"
    force: yes
    mode: 0740
```

This will ensure that if the versions match, and is a config change, to force an overwrite. Have already tested this extensively.

So imo this `force: yes` option is not needed here and will really have no impact,  if we are purging above and unless you have the whole filename specified in the `dest` option: See docs [here](https://docs.ansible.com/ansible/latest/modules/get_url_module.html): 

`If yes and dest is not a directory, will download the file every time and replace the file if the contents change.`

Reply:
Hey @geekbass, I've purged all the upgrade contents form this section here
https://github.com/dcos/dcos-ansible/blob/c1265c42fe1e464512e4b2751a86b87b72871564/roles/DCOS.master/tasks/dcos_upgrade.yml#L1-L4

I do not thing the `force: yes` setting is required anymore as the line listed above will remove the entire directory. 

I am just going to make it true to indicate that this action must happen.

Reply:
Yes you are correct  its not needed anymore @bernadinm however, this purge task is only going to purge directories for config change upgrades and not version upgrades as well. IMO to add a separate purge task just to handle failed config change upgrades is wasteful. Why not purge all older upgrades?

What I propose or suggest is 1 of 2 options. I prefer 1 because I like your introduction of the purge task to keep it clean:

1) Keep the purge task but use it to completely purge the install directory every upgrade and re-create (` path: "{{ path_base_tmp }}/"`). We don' t need these anymore anyways at this point and we are wasting disk space. Nuke all the things. 

In the download task, keep `force: yes` and keep full `dest` like we currently have to overwrite existing if needed.

```
- name: "Upgrade: Purge old directory"
  file:
    path: "{{ path_base_tmp }}/"
    state: absent

- name: "Upgrade: Create download directory"
  file:
    path: "{{ path_base_tmp }}/{{ dcos_version_specifier }}/upgrade_from_{{ ansible_local.dcos_installation['version'] }}/"
    state: directory
    mode: 0755

- name: "Upgrade: Download dcos_node_upgrade.sh"
  get_url:
    url: "{{ dcos['config']['bootstrap_url'] }}/{{ dcos_version_specifier }}/genconf/serve/upgrade_from_{{ ansible_local.dcos_installation['version'] }}/latest/dcos_node_upgrade.sh"
    dest: "{{ path_base_tmp }}/{{ dcos_version_specifier }}/upgrade_from_{{ ansible_local.dcos_installation['version'] }}/dcos_node_upgrade.sh"
    force: yes
    mode: 0740
```

OR

2) Just remove the purge task since it a fix for a one off issue and let the Upgrade task overwrite the file if it exists with `force:yes` and full dest path. 

```
- name: "Upgrade: Download dcos_node_upgrade.sh"
  get_url:
    url: "{{ dcos['config']['bootstrap_url'] }}/{{ dcos_version_specifier }}/genconf/serve/upgrade_from_{{ ansible_local.dcos_installation['version'] }}/latest/dcos_node_upgrade.sh"
    dest: "{{ path_base_tmp }}/{{ dcos_version_specifier }}/upgrade_from_{{ ansible_local.dcos_installation['version'] }}/dcos_node_upgrade.sh"
    force: yes
    mode: 0740
```